src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs

```diff
@@ -159,6 +181,78 @@ private string GetRootPathWithGuid_NoLock()
             return new MetadataAsSourceFile(fileInfo.TemporaryFilePath, navigateLocation, documentName, documentTooltip);
         }
 
+        private async Task<Document> DecompileSymbolAsync(Document temporaryDocument, ISymbol symbol, CancellationToken cancellationToken)
+        {
+            // Get the name of the type the symbol is in
+            var containingOrThis = symbol.GetContainingTypeOrThis();
+            var fullName = GetFullReflectionName(containingOrThis);
+
+            var compilation = await temporaryDocument.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            // TODO: retrieve path to actual assembly instead of reference assembly
+            var reference = compilation.GetMetadataReference(symbol.ContainingAssembly);
+            // Load the assembly.
+            var ad = AssemblyDefinition.ReadAssembly(reference.Display, new ReaderParameters() { AssemblyResolver = new RoslynAssemblyResolver(compilation) });
+
+            // Initialize a decompiler with default settings.
+            var decompiler = new CSharpDecompiler(ad.MainModule, new DecompilerSettings());
+            var fullTypeName = new FullTypeName(fullName);
+
+            // Try to decompile; if an exception is thrown the caller will handle it
+            var text = decompiler.DecompileTypeAsString(fullTypeName);
+            return temporaryDocument.WithText(SourceText.From(text));
+        }
+
+        private class RoslynAssemblyResolver : IAssemblyResolver
+        {
+            private readonly Compilation parentCompilation;
+
+            public RoslynAssemblyResolver(Compilation parentCompilation)
+            {
+                this.parentCompilation = parentCompilation;
+            }
+
+            public AssemblyDefinition Resolve(AssemblyNameReference name)
+            {
+                return Resolve(name, new ReaderParameters());
+            }
+
+            public AssemblyDefinition Resolve(AssemblyNameReference name, ReaderParameters parameters)
+            {
+                foreach (var assembly in parentCompilation.GetReferencedAssemblySymbols())
+                {
+                    if (assembly.Identity.Name == name.Name
+                        && assembly.Identity.Version == name.Version
+                        && assembly.Identity.PublicKeyToken.SequenceEqual(name.PublicKeyToken ?? Array.Empty<byte>()))
+                    {
+                        // reference assemblies should be fine here...
```

Comment:
I admit I don't understand this comment...?

Reply:
The decompiler is fine with metadata-only assemblies for all assemblies referenced by the "main" assembly (i.e. the one that contains the symbol we're trying to decompile). Only the main assembly needs to contain actual IL.